urlfc1c29991b_nil_mish_InJava_tgz-pJ8-edu_usc_cs_nsl_indexing_test_TestTrieProtoJ8/src/edu/usc/cs/nsl/counting/CountingUtil.java

```diff
@@ -0,0 +1,494 @@
+	 * Function to count the words for a given maximum size string in RAM.
+	 * It breaks the count array into chunks. Each chunk size takes into 
+	 * consideration the maximum ram available.
+	 * @param max_string_size maximum string size.
+	 * @return
+	 * @throws IOException
+	 */
+
+	public int countInRAM(int max_string_size) throws IOException{
+		int num_of_chunks = 0;
+		int chunk_size = closestPrime(max_ram/(max_string_size + 8));
+		
+		/* We need to read in long as the file may only have a 
+		 * single alphabet repeated multiple times which will then 
+		 * exceed the INTEGER.MAX_VALUE limit assuming we have a 32 bit JVM.
+		 */		
+		long[] the_big_array = new long[big_array_size];   		
+		BufferedReader br = null;
+		br = new BufferedReader(new FileReader(getTmpFileName(max_string_size)));
+		String the_line = br.readLine();
+		int index = 0;
+		int num_lines = 0;
+		while(the_line != null){
+			index = (int)Math.abs(getHash(the_line.replace("\n", "").replace("\r", ""))%big_array_size);
+			the_big_array[index] += 1;
+			the_line = br.readLine();
+			num_lines++;
+		}
+		br.close();
+		
+		int this_chunk_size = 0;
+		long number_of_words_in_chunk = 0;
+		BufferedWriter bw = new BufferedWriter(new FileWriter(tmpDir+"/chunk_list_"+max_string_size+".txt"));
+		bw.write('0');
+		for(int i = 0; i < big_array_size; i++){
+			if(the_big_array[i] > 0){
+				number_of_words_in_chunk += the_big_array[i]; 
+				this_chunk_size++;
+				if(this_chunk_size == chunk_size - 1){
+					this_chunk_size = 0;
+					num_of_chunks++;
+					bw.write("\t"+(i+1)+"\t"+number_of_words_in_chunk+"\n"+(i+1)); // start of range is inclusive end is exclusive.
+					number_of_words_in_chunk = 0;
+				}
+			}
+		}
+		num_of_chunks++;
+		bw.write("\t"+big_array_size+"\t"+number_of_words_in_chunk+"\n");
+		bw.close();
+		System.out.println("Max String: "+max_string_size+"\tNumLines: "+num_lines+"\tnum of chunks: "+num_of_chunks);
+		return num_of_chunks;
+	}
+	
+	
+	/**
+	 * Function to process chunks. This function uses hashing to do counting.
+	 * Count for words without hash index collision are written to final 
+	 * output file. Words with index collision have the indexes written into 
+	 * a temporary file.
+	 * Note: There are collisions in the RAM counting phase (same big index) 
+	 * and during the processing of chunks (this index). Both these cases
+	 * need to be handled separately.
+	 * 
+	 * @param max_string_size
+	 * @param num_chunks
+	 * @param output_file
+	 * @throws IOException
+	 */
+	
+	public void processChunks(int max_string_size, int num_chunks,
+			String output_file) throws IOException {
+		// The chunks can be loaded in memory and hence they can be counted.
+		int chunk_size = closestPrime(max_ram/(max_string_size + 8));
+		System.out.println("Big array size: "+big_array_size+"\t chunk size: "+chunk_size);
+		BufferedReader chunk_list_file = null;
+		chunk_list_file = new BufferedReader(new FileReader(getTmpChunkListFileName(max_string_size)));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *chunk\_list\_file*. The resource is referenced at lines 249, 250, 257, 266. The resource is closed at line 268. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *chunk\_list\_file* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).